### PR TITLE
gh-115999: Specialize `LOAD_GLOBAL` in free-threaded builds

### PR DESCRIPTION
**Pull request by mpage**
Enable specialization of `LOAD_GLOBAL` in free-threaded builds.  

Thread-safety of specialization in free-threaded builds is provided by the following:

- A critical section is held on both the globals and builtins objects during specialization. This ensures we get an atomic view of both builtins and globals during specialization.
- Generation of new keys versions is made atomic in free-threaded builds.
- Existing helpers are used to atomically modify the opcode.

Thread-safety of specialized instructions in free-threaded builds is provided by the following:

- Relaxed atomics are used when loading and storing dict keys versions. This avoids potential data races as the dict keys versions are read without holding the dictionary's per-object lock in version guards.
- Dicts keys objects are passed from keys version guards to the downstream uops. This ensures that we are loading from the correct offset in the keys object. Once a unicode key has been stored in a keys object for a combined dictionary in free-threaded builds, the offset that it is stored in will never be reused for a different key. Once the version guard passes, we know that we are reading from the correct offset.
- The dictionary read fast-path is used to read values from the dictionary once we know the correct offset.

Performance-wise, this looks like a [3-4% improvement](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241108-3.14.0a1+-0e50451-NOGIL) on free-threaded builds on pyperformance.

* Issue: gh-115999

### CODE REVIEW

**Review by Fidget-Spinner — COMMENTED** [review]
This was less complicated than I initially anticipated. Nice!

LGTM overall, but I'll withold from approval to let others review it first.

File: Objects/dictobject.c [hunk 1]
```diff
@@ -4498,7 +4498,7 @@ dict_popitem_impl(PyDictObject *self)
             return NULL;
         }
     }
-    self->ma_keys->dk_version = 0;
+    FT_ATOMIC_STORE_UINT32_RELAXED(self->ma_keys->dk_version, 0);
```
Comment (colesbury) [inline Objects/dictobject.c:4446]: There's one more assignment to `dk_version` in `insert_split_key` that should use this pattern.

File: Python/specialize.c [hunk 2]
```diff
@@ -1520,104 +1520,103 @@ PyObject *descr, DescriptorClassification kind, bool is_method)
     return 1;
 }
 
-void
-_Py_Specialize_LoadGlobal(
+static void
+specialize_load_global_lock_held(
```
Comment (colesbury) [inline Python/specialize.c:1523]: We should mark the dicts as "shared" (`_PyObject_GC_SET_SHARED` / `SET_DICT_SHARED`) when we specialize for load global. That ensures that the keys object will not be immediately freed when the dictionary is resized, which is important if another thread is trying to read the keys without holding a lock.

The dict code enforces this with `ensure_shared_on_read()`, but I think we should have the version check subsume this: if the dict has a valid version, then it's marked as shared (and we can attempt the non-locking read without checking the owning thread).

So maybe ensure the dict is shared in `_PyDictKeys_GetVersionForCurrentState`.

File: Python/bytecodes.c [hunk 3]
```diff
@@ -1599,47 +1599,70 @@ dummy_func(
...
             SYNC_SP();
             DEOPT_IF(res_o == NULL);
+            #if Py_GIL_DISABLED
+            int increfed = _Py_TryIncrefCompare(&entries[index].me_value, res_o);
+            DEOPT_IF(!increfed);
+            PyDictObject *dict = (PyDictObject*) GLOBALS();
+            if (globals_keys != _Py_atomic_load_ptr_acquire(&dict->ma_keys)) {
+                Py_DECREF(res_o);
+                DEOPT_IF(true);
+            }
```
Comment (colesbury) [inline Python/bytecodes.c]: I *think* these checks that the keys haven't changed are not necessary. We have them in `dictobject.c` too, and I think they're not necessary there either.

I think the checks were inherited from the nogil forks, which used a different scheme for dictionary keys.

(Also the use of `global_keys` here seems to contradict the `DEAD(globals_keys)` above.)

File: Python/bytecodes.c [hunk 4]
```diff
@@ -1599,47 +1599,70 @@ dummy_func(
...
 
         op(_LOAD_GLOBAL_MODULE_FROM_KEYS, (index/1, globals_keys: PyDictKeysObject* -- res, null if (oparg & 1))) {
             PyDictUnicodeEntry *entries = DK_UNICODE_ENTRIES(globals_keys);
-            PyObject *res_o = entries[index].me_value;
+            PyObject *res_o = FT_ATOMIC_LOAD_PTR_RELAXED(entries[index].me_value);
             DEAD(globals_keys);
             SYNC_SP();
             DEOPT_IF(res_o == NULL);
+            #if Py_GIL_DISABLED
+            int increfed = _Py_TryIncrefCompare(&entries[index].me_value, res_o);
```
Comment (colesbury) [inline Python/bytecodes.c]: I think we want to support deferred reference counting here. Loading functions to call them will use `LOAD_GLOBAL_MODULE` and `LOAD_GLOBAL_BUILTINS`.

I don't think we have the equivalent of `_Py_TryIncrefCompare` for `_PyStackRef` yet, it needs to be written.

**Review by colesbury — COMMENTED** [review]
Nice! A few comments below.

File: Objects/dictobject.c [hunk 5]
```diff
@@ -1285,6 +1285,20 @@ ensure_shared_on_resize(PyDictObject *mp)
 #endif
 }
 
+static inline void
+ensure_shared_on_keys_version_assignment(PyDictObject *mp)
+{
+    ASSERT_DICT_LOCKED((PyObject *) mp);
+    #ifdef Py_GIL_DISABLED
+    if (!_Py_IsOwnedByCurrentThread((PyObject *)mp) && !IS_DICT_SHARED(mp)) {
```
Comment (colesbury) [inline Objects/dictobject.c]: I think this check should just be: `if (!IS_DICT_SHARED(mp))`. 

The `_Py_IsOwnedByCurrentThread()` check here might be okay with some other changes and a few assumptions (like specialization remains per-thread), but I think it's easier to reason about if this is just marks the dict as shared if it's not already.

File: Objects/dictobject.c [hunk 6]
```diff
@@ -7313,20 +7327,67 @@ _PyDictKeys_DecRef(PyDictKeysObject *keys)
...
+    return res.version;
+}
+
+uint32_t
+_PyDict_GetKeysVersionForCurrentState(PyInterpreterState *interp,
+                                      PyDictObject *dict)
+{
+    ASSERT_DICT_LOCKED((PyObject *) dict);
+    assign_keys_version_result res = assign_keys_version(interp, dict->ma_keys);
+    if (res.assigned) {
```
Comment (colesbury) [inline Objects/dictobject.c]: I don't think you need the `res.assigned` check and the combination with `_Py_IsOwnedByCurrentThread` in `ensure_shared_on_keys_version_assignment` is not correct, I think: we could end up not marking the dict as shared in a non-owning thread.

Marking the dict as shared is cheap and this is not called frequently, and I expect the code will be simpler without the `assign_keys_version_result`.

File: Include/internal/pycore_dict.h [hunk 7]
```diff
@@ -90,6 +90,15 @@ extern PyObject *_PyDict_FromKeys(PyObject *, PyObject *, PyObject *);
...
         PyInterpreterState *interp, PyDictKeysObject *dictkeys);
 
+/* Gets a version number unique to the current state of the keys of dict, if possible.
+ *
+ * In free-threaded builds ensures that the dict can be used for lock-free
+ * reads if a version was assigned.
+ *
+ * Returns the version number, or zero if it was not possible to get a version number. */
+extern uint32_t _PyDict_GetKeysVersionForCurrentState(
+        PyInterpreterState *interp, PyDictObject *dictkeys);
```
Comment (colesbury) [inline Include/internal/pycore_dict.h]: s/dictkeys/dict

File: Objects/dictobject.c (hunk 5 above)
Comment (mpage) [reply Objects/dictobject.c]: 🤦 yep

**Review by colesbury — APPROVED** [review]

**Comment by mpage** [comment]
!buildbot nogil refleak

**Pull request merged**